### PR TITLE
FIx #163 - Array2d.* don't work on non-zero-based arrays

### LINKED ISSUE #163: Array slicing does not work properly with non 0-based arrays
opened at [CodePlex](https://visualfsharp.codeplex.com/workitem/181) by [porges](https://www.codeplex.com/site/users/view/Porges)

Despite the rest of the Array2D methods working correctly:
```fsharp
(Array2D.createBased 5 4 3 2 1).[*, *] // boom!
```

### PR DESCRIPTION
**Pull request by dsyme**
Spec clarification:

* Slicing ``arr.[slice]`` from a non-zero based multi-dimensional array gives a _zero_-based multi-dimensional array

* Assigning ``dst.[slice] <- src`` into a slice of a non-zero based multi-dimensional array requires that ``src`` be a _zero_-based multi-dimensional array on the right of the assignment 

Tests included.

### CODE REVIEW

File: tests/fsharp/core/array/test.fsx [hunk 1]
```diff
@@ -594,14 +594,63 @@ module Array2Tests = begin
...
 
   let _ = test_make_get_set_length ()
 
+
 end
 
+module ArrayNonZeroBasedTestsSlice = 
+  let runTest () = 
+    let arr = (Array2D.initBased 5 4 3 2 (fun i j -> (i,j)))
+    test "fewoih1" (arr.[6,*] = [|(6, 4); (6, 5)|])
```
Comment (forki) [inline tests/fsharp/core/array/test.fsx]: I always wonder what these magic strings mean. Is that the test name?

Comment (dsyme) [inline tests/fsharp/core/array/test.fsx]: DFUIDs (Don's Finger-Unique Identifiers)

Comment (enricosada) [inline tests/fsharp/core/array/test.fsx]: :smile: i was thinking about it too

File: tests/fsharp/core/array/test.fsx [hunk 2]
```diff
@@ -594,14 +594,63 @@ module Array2Tests = begin
...
+    let arr2d = 
+        let arr = Array2D.zeroCreateBased 5 4 3 2 
+        for i in 5..7 do for j in 4..5 do arr.[i,j] <- (i,j)
+        arr
+    let arr2d2 = 
+        let arr = Array2D.zeroCreate 3 2 
+        for i in 0..2 do for j in 0..1 do arr.[i,j] <- (j,i)
+        arr
+    test "fewoih11" (arr2d.[6..6,5] =  [|(6, 5)|])
+    test "fewoih11" (arr2d.[..6,5] =  [|(6, 5)|])
```
Comment (latkin) [inline tests/fsharp/core/array/test.fsx]: This guy fails, actual result is `[|(5, 5); (6, 5)|]`, which is correct.

File: tests/fsharp/core/array/test.fsx [hunk 3]
```diff
@@ -594,14 +594,63 @@ module Array2Tests = begin
...
     Array2D.blit b 0 0 a 0 0 2 2
     //test "a2_blit"
     //      (Array2D.sub a 0 0 2 2 = b)
 
   let _ = test_make_get_set_length ()
 
+
 end
 
+module ArrayNonZeroBasedTestsSlice =
```
Comment (latkin) [inline tests/fsharp/core/array/test.fsx]: The non-zero-based stuff is all missing from portable, need to `#if` this out

**Pull request closed without merging**